dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,807 @@
+            throw Exception("Too many arguments", ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION);
+
+        bool has_nullable_col = false;
+        bool has_data_type_nothing = false;
+
+        for (const auto & arg : arguments)
+            checkInputArg(arg, true, &has_nullable_col, &has_data_type_nothing);
+
+        if (has_data_type_nothing)
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNothing>());
+
+        if (has_nullable_col)
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNumber<ResultType>>());
+        else
+            return std::make_shared<DataTypeNumber<ResultType>>();
+    }
+
+    template <typename ExprT, typename PatT, typename MatchTypeT>
+    void REGEXP_CLASS_MEM_FUNC_IMPL_NAME(ColumnWithTypeAndName & res_arg, const ExprT & expr_param, const PatT & pat_param, const MatchTypeT & match_type_param) const
+    {
+        size_t col_size = expr_param.getDataNum();
+
+        // Check if args are all const columns
+        if constexpr (ExprT::isConst() && PatT::isConst() && MatchTypeT::isConst())
+        {
+            if (expr_param.isNullAt(0) || pat_param.isNullAt(0) || match_type_param.isNullAt(0))
+            {
+                res_arg.column = res_arg.type->createColumnConst(col_size, Null());
+                return;
+            }
+
+            int flags = getDefaultFlags();
+            String expr = expr_param.getString(0);
+            String pat = pat_param.getString(0);
+            if (unlikely(pat.empty()))
+                throw Exception(EMPTY_PAT_ERR_MSG);
+
+            String match_type = match_type_param.getString(0);
+
+            Regexps::Regexp regexp(addMatchTypeForPattern(pat, match_type, collator), flags);
+            ResultType res{regexp.match(expr)};
+            res_arg.column = res_arg.type->createColumnConst(col_size, toField(res));
+            return;
+        }
+
+        // Initialize result column
+        auto col_res = ColumnVector<ResultType>::create();
+        typename ColumnVector<ResultType>::Container & vec_res = col_res->getData();
+        vec_res.resize(col_size, 0);
+
+        constexpr bool has_nullable_col = ExprT::isNullableCol() || PatT::isNullableCol() || MatchTypeT::isNullableCol();
+
+        // Start to match
+        if constexpr (canMemorize<PatT, MatchTypeT>())
+        {
+            std::unique_ptr<Regexps::Regexp> regexp;
+            if (col_size > 0)
+            {
+                regexp = memorize(pat_param, match_type_param, collator);
+                if (regexp == nullptr)
+                {
+                    auto nullmap_col = ColumnUInt8::create();
+                    typename ColumnUInt8::Container & nullmap = nullmap_col->getData();
+                    nullmap.resize(col_size, 1);
+                    res_arg.column = ColumnNullable::create(std::move(col_res), std::move(nullmap_col));
+                    return;
+                }
+            }
```

Comment:
how about adding an `else` branch to return directly if `col_size == 0`?

Reply:
> how about adding an `else` branch to return directly if `col_size == 0`?

I think it's unnecessary, as we will add more duplicated codes.